lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -922,8 +922,43 @@ WitnessChecker::lookupValueWitnessesViaImplementsAttr(
   }
 }
 
-SmallVector<ValueDecl *, 4> 
-WitnessChecker::lookupValueWitnesses(ValueDecl *req, bool *ignoringNames) {
+bool WitnessChecker::isMemberOperator(FuncDecl *decl, Type type) {
+  auto *DC = decl->getDeclContext();
+  auto selfNominal = DC->getSelfNominalTypeDecl();
+
+  // Check the parameters for a reference to 'Self'.
+  bool isProtocol =
+      selfNominal && isa<ProtocolDecl>(selfNominal) &&
+      TC.conformsToProtocol(type, dyn_cast<ProtocolDecl>(selfNominal), DC,
+                            ConformanceCheckFlags::InExpression);
+
+  for (auto param : *decl->getParameters()) {
+    auto paramType = param->getInterfaceType();
+    if (!paramType)
+      break;
+
+    // Look through a metatype reference, if there is one.
+    paramType = paramType->getMetatypeInstanceType();
+
+    // Is it the same nominal type?
+    auto nominal = paramType->getAnyNominal();
+    if ((!nominal && paramType->getKind() != TypeKind::Tuple) ||
```

Comment:
I’m not sure I understand the check here. Other than nominals, what are you accepting here? The ‘Self’ generic parameter? Then you might want to make this more precise and check for a GenericTypeParamType. Also as a general stylistic comment, instead of checking getKind() it’s better to use ->is<TupleType>(), etc.

Reply:
No, I'm trying to check for generic parameters that might fulfill the requirement:

```swift
infix operator <~>
protocol P1 {
	static func <~>(x: Self, y: Self)
}

protocol P2 {}

struct ConformsWithMoreGenericWitnesses : P1, P2 {
}
func <~> <P: P2>(x: P, y: P) {}
```

It looks like `is<GenericTypeParamType>()` does what I want. I couldn't find the right way to express that.